src/GraphQL/Caching/DocumentCache.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using GraphQL.Language.AST;
+
+namespace GraphQL.Caching
+{
+    /// <summary>
+    /// The default implementation of a document cache, limited by a configured amount of memory.
+    /// </summary>
+    public class DocumentCache : IDocumentCache
```

Comment:
I don't think it makes sense to invent your own implementations. It is enough to delegate all the work to `IMemoryCache`. It will be easier, less error prone and more customizable. I suggest to use 
```c#
public sealed class DefaultDocumentCache : IDocumentCache
{
  public Document this[string query] { get => null; } { set => { } }
}
```
and get rid of some null checks in `DocumentExecuter` and alse pass `new DefaultDocumentCache()` into `DocumentExecuter` ctor (or even static single instance since default impl is stateless), not null + change ` _documentCache = documentCache;` to ` _documentCache = documentCache ?? throw new ArgumentNullException(nameof(documentCache));;

Reply:
Of course `MemoryDocumentCache` should be placed in new project since it depends on custom implementation from `Microsoft.Extensions.Caching.Memory`

Reply:
There’s no dependency on caching right now, is there? It seems a shame to add another dependency for such a simple implementation.

Another idea: We could move the caching implementation to a separate library and add a dependency there. 

I’ll make all the rest of your suggestions.

Reply:
> Another idea: We could move the caching implementation to a separate library and add a dependency there.

That's what I wanted.

Reply:
And _default_ implementation is just "No caching".

Reply:
Done